Implement ReadableStreamDefaultReader

Taym95 opened the pull request:
* ReleaseLock
* Closed
* Cancel

<!-- Please describe your changes on the following line: -->
Part of #32898

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

<!-- Either: -->

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Taym95 commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
     }
 
     /// <https://streams.spec.whatwg.org/#default-reader-release-lock>
     fn ReleaseLock(&self) {
```
@gterzian @wusyong this can be reviewed.

pewsheen commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
-        // TODO: use TypeError.
         // <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
+        // step 3.2
+        self.empty_read_requests();
+        // step 3.3
         for request in self.read_requests.borrow_mut().drain(0..) {
-            request.error_steps();
+            request.error_steps(error.clone());
```
I think we should clone `self.read_requests` somewhere before clearing it. So that on the step 3.3 we can perform `error_steps()` to those `read_requests`

pewsheen replied:
or maybe we can do `drain` only?

Taym95 replied:
indeed thanks

gterzian commented on components/script/dom/readablestream.rs:
```diff
@@ -213,6 +213,14 @@ impl ReadableStream {
         }
     }
 
+    // Call into the rellease steps of the controller,
```
typo

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
+    fn Cancel(&self, _cx: SafeJSContext, reason: SafeHandleValue) -> Rc<Promise> {
+        let promise = Promise::new(&self.reflector_.global());
+        // step 1
+        if self.stream.get().is_none() {
```
I think it would be clearer to invert the logic, by doing `if let Some(stream) =`, and doing step 1 in the `else`.

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
+            return;
+        }
+
+        // <https://streams.spec.whatwg.org/#readable-stream-reader-generic-release>
```
let's put this in its own method, it is used in many other places in the spec, and also it's clearer with the spec link. We can do the same with other spec functions.

gterzian reviewed (changes requested):
Thank you! I think a good next step would be to break-down the algos in their own methods where appropriate.

Taym95 commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,93 @@ impl ReadableStreamDefaultReader {
+        stream.perform_release_steps();
+
+        // step 2.1.8 & 2.1.9
+        self.stream.set(None);
```
Yeah I wanted to do that, but when we set `stream` to `None` then`reader` will be dropped so it is also undefined?

gterzian commented on components/script/dom/readablestream.rs:
```diff
@@ -423,6 +431,47 @@ impl ReadableStream {
+        closed_promise: &Promise,
+        read_requests: DomRefCell<VecDeque<ReadRequest>>,
+    ) {
+        // step 2.1.3
```
I think you can start counting from 1, since you are in the generic algo.

gterzian commented on components/script/dom/readablestream.rs:
```diff
@@ -423,6 +431,47 @@ impl ReadableStream {
+        self.set(None);
+
+        // step  3.1
+        let error = Error::Type("No chunks are available because the stream is errored".to_owned());
```
This can be moved to a method implementing https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests on the reader.  You can use the code at https://streams.spec.whatwg.org/#readable-stream-error and move it to a new method called in both places.

gterzian commented on components/script/dom/readablestream.rs:
```diff
@@ -423,6 +431,47 @@ impl ReadableStream {
+        // step 3.2
+        read_requests.borrow_mut().clear();
+        // step 3.3
+        for request in read_requests_cloned.borrow_mut().drain(0..) {
```
I don't think you need the two steps above, and here you should do something like [here](https://github.com/servo/servo/blob/1ff141d486ddd0f28c1af45c9f873154138c9b47/components/script/dom/readablestreamdefaultreader.rs#L147), because the error steps can call back into the stream and cause a borrow error(but see above comment, you should re-use the existing code anyway).

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -74,10 +75,10 @@ impl ReadRequest {
-    pub fn error_steps(&self) {
+    pub fn error_steps(&self, error: Error) {
         match self {
             // TODO: pass error type.
```
todo can be removed.

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -161,11 +161,11 @@ impl ReadableStreamDefaultReader {
-    pub fn error(&self, _error: Error) {
+    pub fn error(&self, error: Error) {
         self.closed_promise.reject_native(&());
         let pending_requests = mem::take(&mut *self.read_requests.borrow_mut());
```
this part can be put in a separate method and called elsewhere(see above)

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -176,43 +176,57 @@ impl ReadableStreamDefaultReader {
     }
+
+    /// <https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreadererrorreadrequests>
+    pub fn empty_read_requests(&self) {
```
this seems unused, and perhaps not necessary(see other comments).

gterzian reviewed (changes requested):
Looks good, with some changes suggested.

Taym95 committed:
Implement ReadableStreamDefaultReader

 Make the stream mutable
 readable-stream-reader-generic-release
 Proper error types when releasing
 Closed
 Cancel

Signed-off-by: Taym <haddadi.taym@gmail.com>

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -169,45 +165,81 @@ impl ReadableStreamDefaultReader {
-        self.stream
-            .perform_pull_steps(ReadRequest::Read(promise.clone()));
+        if let Some(stream) = self.stream.get() {
+            stream.perform_pull_steps(ReadRequest::Read(promise.clone()));
```
This should call into https://streams.spec.whatwg.org/#readable-stream-default-reader-read (we don't have that method yet).

Taym95 replied:
I needed to changed this code because I converted stream to `MutNullableDom<ReadableStream>`, this `read` function is not implemented correctly, I think we can fix it in next PR when we add https://streams.spec.whatwg.org/#readable-stream-default-reader-read

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -169,45 +165,81 @@ impl ReadableStreamDefaultReader {
+                request.error_steps(rval.handle());
+            }
+        } else {
+            // step 1
```
I think we should also follow the spec more closely:

- add a conditional on `self.stream.get().is_none`().
- Call into a new method for https://streams.spec.whatwg.org/#abstract-opdef-readablestreamdefaultreaderrelease
- Move rest of algo to that new method(it can be called from other places later).

gterzian reviewed (changes requested):
Looks good, with a couple of points.

Taym95 commented on components/script/dom/readablestream.rs:
```diff
@@ -482,6 +489,35 @@ impl ReadableStream {
+            closed_promise.reject_error(Error::Type("stream state is not readable".to_owned()));
+        } else {
+            // step 4
+            closed_promise.reject_error(Error::Type(
```
indeed I missed this:

Otherwise, set reader.[[[closedPromise]]](https://streams.spec.whatwg.org/#readablestreamgenericreader-closedpromise) to [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a [TypeError](https://tc39.es/ecma262/#sec-native-error-types-used-in-this-standard-typeerror) exception.

thanks

Taym95 committed:
follow the spec more closely

Signed-off-by: Taym <haddadi.taym@gmail.com>

Taym95 commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -169,45 +165,135 @@ impl ReadableStreamDefaultReader {
+            return;
+        } else {
+            // step 2
+            self.release();
```
@gterzian I tired to follow the spec more closely, please check again, thanks

gterzian reviewed (approved):
LGTM, thanks!
